### PR TITLE
feat: @requires directive on different services as gateway

### PR DESCRIPTION
**Pull request by simone-sanfratello**
Following https://github.com/mercurius-js/mercurius/issues/456, it provide support for `@requires` directive to work properly on gateway on different services

---

Using benchmarks from https://github.com/mercurius-js/auth/tree/main/bench

**this branch**

```
==============================
= Normal Mode                =
==============================
Running 10s test @ http://localhost:3000/graphql
100 connections

┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼───────┤
│ Latency │ 4 ms │ 4 ms │ 6 ms  │ 11 ms │ 4.58 ms │ 2.26 ms │ 89 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬──────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg      │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┼─────────┤
│ Req/Sec   │ 11231   │ 11231   │ 20079   │ 20319   │ 19160.37 │ 2523.42 │ 11228   │
├───────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┼─────────┤
│ Bytes/Sec │ 5.91 MB │ 5.91 MB │ 10.6 MB │ 10.7 MB │ 10.1 MB  │ 1.33 MB │ 5.91 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴──────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

211k requests in 11.03s, 111 MB read
===============================
= Gateway Mode                =
===============================
Running 10s test @ http://localhost:3000/graphql
100 connections

┌─────────┬───────┬───────┬───────┬───────┬──────────┬──────────┬────────┐
│ Stat    │ 2.5%  │ 50%   │ 97.5% │ 99%   │ Avg      │ Stdev    │ Max    │
├─────────┼───────┼───────┼───────┼───────┼──────────┼──────────┼────────┤
│ Latency │ 27 ms │ 28 ms │ 55 ms │ 73 ms │ 30.45 ms │ 13.14 ms │ 269 ms │
└─────────┴───────┴───────┴───────┴───────┴──────────┴──────────┴────────┘
┌───────────┬────────┬────────┬─────────┬─────────┬─────────┬────────┬────────┐
│ Stat      │ 1%     │ 2.5%   │ 50%     │ 97.5%   │ Avg     │ Stdev  │ Min    │
├───────────┼────────┼────────┼─────────┼─────────┼─────────┼────────┼────────┤
│ Req/Sec   │ 1284   │ 1284   │ 3509    │ 3577    │ 3229.1  │ 658.7  │ 1284   │
├───────────┼────────┼────────┼─────────┼─────────┼─────────┼────────┼────────┤
│ Bytes/Sec │ 450 kB │ 450 kB │ 1.23 MB │ 1.25 MB │ 1.13 MB │ 231 kB │ 449 kB │
└───────────┴────────┴────────┴─────────┴─────────┴─────────┴────────┴────────┘

Req/Bytes counts sampled once per second.

36k requests in 11.02s, 12.4 MB read
```

**master branch**

```
==============================
= Normal Mode                =
==============================
Running 10s test @ http://localhost:3000/graphql
100 connections

┌─────────┬──────┬──────┬───────┬───────┬─────────┬────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev  │ Max   │
├─────────┼──────┼──────┼───────┼───────┼─────────┼────────┼───────┤
│ Latency │ 4 ms │ 4 ms │ 6 ms  │ 13 ms │ 4.58 ms │ 2.3 ms │ 89 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬──────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg      │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┼─────────┤
│ Req/Sec   │ 11159   │ 11159   │ 19983   │ 20687   │ 19107.28 │ 2585.38 │ 11157   │
├───────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┼─────────┤
│ Bytes/Sec │ 5.87 MB │ 5.87 MB │ 10.5 MB │ 10.9 MB │ 10 MB    │ 1.36 MB │ 5.87 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴──────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.

210k requests in 11.02s, 111 MB read
===============================
= Gateway Mode                =
===============================
Running 10s test @ http://localhost:3000/graphql
100 connections

┌─────────┬───────┬───────┬───────┬───────┬──────────┬──────────┬────────┐
│ Stat    │ 2.5%  │ 50%   │ 97.5% │ 99%   │ Avg      │ Stdev    │ Max    │
├─────────┼───────┼───────┼───────┼───────┼──────────┼──────────┼────────┤
│ Latency │ 26 ms │ 28 ms │ 53 ms │ 76 ms │ 30.82 ms │ 15.51 ms │ 309 ms │
└─────────┴───────┴───────┴───────┴───────┴──────────┴──────────┴────────┘
┌───────────┬────────┬────────┬─────────┬─────────┬─────────┬────────┬────────┐
│ Stat      │ 1%     │ 2.5%   │ 50%     │ 97.5%   │ Avg     │ Stdev  │ Min    │
├───────────┼────────┼────────┼─────────┼─────────┼─────────┼────────┼────────┤
│ Req/Sec   │ 1290   │ 1290   │ 3443    │ 3617    │ 3190.9  │ 673.72 │ 1290   │
├───────────┼────────┼────────┼─────────┼─────────┼─────────┼────────┼────────┤
│ Bytes/Sec │ 452 kB │ 452 kB │ 1.21 MB │ 1.27 MB │ 1.12 MB │ 236 kB │ 452 kB │
└───────────┴────────┴────────┴─────────┴─────────┴─────────┴────────┴────────┘

Req/Bytes counts sampled once per second.

32k requests in 10.03s, 11.2 MB read
```

### CODE REVIEW

File: test/gateway/request.js [hunk 1]
```diff
@@ -8,13 +8,7 @@ test('sendRequest method rejects when request errs', t => {
...
     method: 'POST',
     body: JSON.stringify({
-      query: `
-      query ServiceInfo {
-        _service {
-          sdl
-        }
-      }
-      `
+      query: 'invalid query'
```
Comment (simoneb) [inline test/gateway/request.js:11]: why this change? it doesn't seem related to the feature

**Review by simoneb — COMMENTED** [review]
A few minor comments but overall LGTM. Good stuff!

File: lib/federation.js [hunk 2]
```diff
@@ -128,10 +128,12 @@ function getStubTypes (schemaDefinitions, isGateway) {
 
 function gatherDirectives (type) {
   let directives = []
-  for (const node of (type.extensionASTNodes || [])) {
+  if (type.extensionASTNodes) {
```
[the remark this answers is not in the document]
Comment (simone-sanfratello) [reply lib/federation.js:131]: perfomance

File: lib/gateway/service-map.js [hunk 3]
```diff
@@ -15,8 +15,18 @@ const { MER_ERR_GQL_GATEWAY_INIT } = require('../errors')
 const buildFederationSchema = require('../federation')
 
 function hasDirective (directiveName, node) {
-  const { directives = [] } = node
-  return directives.some(directive => directive.name.value === directiveName)
```
[the remark this answers is not in the document]
Comment (simone-sanfratello) [reply lib/gateway/service-map.js:19]: performance

File: lib/federation.js [hunk 4]
```diff
@@ -81,11 +81,11 @@ function removeExternalFields (typeDefinition) {
   }
 }
 
-function hasExtendsDirective (definition) {
+function hasExtensionDirective (definition) {
```
[the remark this answers is not in the document]
Comment (simone-sanfratello) [reply lib/federation.js:84]: we are looking for one of the directives `requires` or `extends`

File: lib/gateway/make-resolver.js [hunk 5]
```diff
@@ -38,7 +38,32 @@ function getDirectiveSelection (node, directiveName) {
...
+
+  const requires = []
+  const selectedFields = selections.map(selection => selection.name.value)
+
+  for (const node of type.extensionASTNodes) {
+    for (const field of node.fields) {
+      if (!selectedFields.includes(field.name.value)) {
+        continue
+      }
+      const directive = field.directives.find(d => d.name.value === 'requires')
```
Comment (mcollina) [inline lib/gateway/make-resolver.js:56]: can this be null for some reason?

File: lib/federation.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (simoneb) [reply lib/federation.js:125]: not really, just curious as it doesn't stick out to me as a performance improvement :)

File: lib/gateway/service-map.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (simoneb) [reply lib/gateway/service-map.js:19]: same as above, it's fine just curious as it doesn't look like an (obvious) performance improvement

File: lib/federation.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (simone-sanfratello) [reply lib/federation.js:125]: in case `type.extensionASTNodes` is falsy, it's creating a new array to not iterate it, btw, should I remove it?

[the remark this answers is not in the document]
Comment (simoneb) [reply lib/federation.js:125]: no you're fine!

File: lib/gateway/make-resolver.js (hunk 5 above)
Comment (simone-sanfratello) [reply lib/gateway/make-resolver.js:57]: I'm writing a test case

File: bench/gateway-without-auth.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
+'use strict'
+
```
Comment (mcollina) [inline bench/gateway-without-auth.js:2]: Please rename this to just "gateway"

**Review by mcollina — COMMENTED** [review]
This implementation looks good to me. Waiting for @PabloSzx @PacoDu @jonnydgreen to take a look.

**Review by jonnydgreen — COMMENTED** [review]
Looks good to me, nothing more to add than what's already been said!

File: lib/gateway/make-resolver.js (hunk 5 above)
Comment (simone-sanfratello) [reply lib/gateway/make-resolver.js:57]: can't find a case. help?

Comment (jonnydgreen) [reply lib/gateway/make-resolver.js:57]: Looking at the logic, any non-external field containing a directive that isn't `requires` should get you the test case.

For example:

Schema:
```graphql
directive @custom on OBJECT | FIELD_DEFINITION

type Size @key(fields: "id") {
  id: Int!
  cpus: Int!
  memory: Int!
}

type Region @key(fields: "id") {
  id: Int!
  city: String!
}

type Metadata @key(fields: "id") {
  id: Int!
  description: String!
}

extend type Host @key(fields: "id") {
  id: Int! @external
  size: Int! @external
  region: Int! @external
  # Field defined here with directive that isn't `requires`
  metadata: Metadata! @custom
  sizeData: Size! @requires(fields: "size")
  regionData: Region! @requires(fields: "region")
}
```

Query:
```graphql
query {
  hosts {
    name
    sizeData {
      cpus
    }
    metadata {
      description
    }
  }
}
```

File: bench/normal.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+'use strict'
+
```
Comment (mcollina) [inline bench/normal.js:2]: I would call it standalone.js

File: lib/util.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  for (let i = 0; i < node.directives.length; i++) {
+    if (node.directives[i].name.value === directiveName) {
+      return true
+    }
+  }
+}
+
+function hasExtensionDirective (node) {
+  if (!node.directives || node.directives.length < 1) {
+    return
```
Comment (mcollina) [inline lib/util.js:16]:
```suggestion
    return false
```

File: lib/util.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
+'use strict'
+
+function hasDirective (directiveName, node) {
+  if (!node.directives || node.directives.length < 1) {
+    return
```
Comment (mcollina) [inline lib/util.js:5]:
```suggestion
    return false
```

File: lib/gateway/make-resolver.js (hunk 5 above)
Comment (simone-sanfratello) [reply lib/gateway/make-resolver.js:57]: many thanks @jonnydgreen !

**Review by mcollina — APPROVED** [review]
lgtm

**Review by PabloSzx — APPROVED** [review]

**Review by jonnydgreen — APPROVED** [review]
LGTM!

**Review by PacoDu — APPROVED** [review]
LGTM

tho I do not understand why the windows test CI does not reach 100% coverage but the linux one does
```
ERROR: Coverage for lines (98.88%) does not meet global threshold (100%)
ERROR: Coverage for functions (99.19%) does not meet global threshold (100%)
ERROR: Coverage for branches (98.82%) does not meet global threshold (100%)
ERROR: Coverage for statements (98.91%) does not meet global threshold (100%)
--------------------------------------|---------|----------|---------|---------|-------------------
File                                  | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
--------------------------------------|---------|----------|---------|---------|-------------------
All files                             |   98.91 |    98.82 |   99.19 |   98.88 |                   
 mercurius                            |   99.65 |    98.51 |     100 |   99.65 |                   
  index.js                            |   99.65 |    98.51 |     100 |   99.65 | 197               
 mercurius/lib                        |   99.41 |    99.28 |   98.88 |   99.39 |                   
  errors.js                           |     100 |      100 |     100 |     100 |                   
  federation.js                       |     100 |      100 |     100 |     100 |                   
  gateway.js                          |   96.38 |    96.61 |   92.59 |   96.24 | 357,364,367-369   
  handlers.js                         |     100 |      100 |     100 |     100 |                   
  hooks.js                            |     100 |    97.73 |     100 |     100 | 52                
  persistedQueryDefaults.js           |     100 |      100 |     100 |     100 |                   
  queryDepth.js                       |     100 |      100 |     100 |     100 |                   
  routes.js                           |     100 |      100 |     100 |     100 |                   
  subscriber.js                       |     100 |      100 |     100 |     100 |                   
  subscription-client.js              |     100 |      100 |     100 |     100 |                   
  subscription-connection.js          |     100 |      100 |     100 |     100 |                   
  subscription-protocol.js            |     100 |      100 |     100 |     100 |                   
  subscription.js                     |     100 |      100 |     100 |     100 |                   
  symbols.js                          |     100 |      100 |     100 |     100 |                   
  util.js                             |     100 |      100 |     100 |     100 |                   
 mercurius/lib/federation             |     100 |      100 |     100 |     100 |                   
  compositionRules.js                 |     100 |      100 |     100 |     100 |                   
 mercurius/lib/gateway                |   96.77 |     97.9 |     100 |   96.61 |                   
  make-resolver.js                    |     100 |      100 |     100 |     100 |                   
  request.js                          |     100 |      100 |     100 |     100 |                   
  service-map.js                      |   90.91 |    91.18 |     100 |   90.91 | 92-97,141-147     
 mercurius/tap-snapshots/test         |     100 |      100 |     100 |     100 |                   
  errors.js.test.cjs                  |     100 |      100 |     100 |     100 |                   
  federation.js.test.cjs              |     100 |      100 |     100 |     100 |                   
  reply-decorator.js.test.cjs         |     100 |      100 |     100 |     100 |                   
  routes.js.test.cjs                  |     100 |      100 |     100 |     100 |                   
 mercurius/tap-snapshots/test/gateway |     100 |      100 |     100 |     100 |                   
  remote-services.js.test.cjs         |     100 |      100 |     100 |     100 |                   
--------------------------------------|---------|----------|---------|---------|-------------------
```

**Comment by PabloSzx** [comment]
> LGTM
> 
> tho I do not understand why the windows test CI does not reach 100% coverage but the linux one does
> 
> ```
> ERROR: Coverage for lines (98.88%) does not meet global threshold (100%)
> ERROR: Coverage for functions (99.19%) does not meet global threshold (100%)
> ERROR: Coverage for branches (98.82%) does not meet global threshold (100%)
> ERROR: Coverage for statements (98.91%) does not meet global threshold (100%)
> --------------------------------------|---------|----------|---------|---------|-------------------
> File                                  | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
> --------------------------------------|---------|----------|---------|---------|-------------------
> All files                             |   98.91 |    98.82 |   99.19 |   98.88 |                   
>  mercurius                            |   99.65 |    98.51 |     100 |   99.65 |                   
>   index.js                            |   99.65 |    98.51 |     100 |   99.65 | 197               
>  mercurius/lib                        |   99.41 |    99.28 |   98.88 |   99.39 |                   
>   errors.js                           |     100 |      100 |     100 |     100 |                   
>   federation.js                       |     100 |      100 |     100 |     100 |                   
>   gateway.js                          |   96.38 |    96.61 |   92.59 |   96.24 | 357,364,367-369   
>   handlers.js                         |     100 |      100 |     100 |     100 |                   
>   hooks.js                            |     100 |    97.73 |     100 |     100 | 52                
>   persistedQueryDefaults.js           |     100 |      100 |     100 |     100 |                   
>   queryDepth.js                       |     100 |      100 |     100 |     100 |                   
>   routes.js                           |     100 |      100 |     100 |     100 |                   
>   subscriber.js                       |     100 |      100 |     100 |     100 |                   
>   subscription-client.js              |     100 |      100 |     100 |     100 |                   
>   subscription-connection.js          |     100 |      100 |     100 |     100 |                   
>   subscription-protocol.js            |     100 |      100 |     100 |     100 |                   
>   subscription.js                     |     100 |      100 |     100 |     100 |                   
>   symbols.js                          |     100 |      100 |     100 |     100 |                   
>   util.js                             |     100 |      100 |     100 |     100 |                   
>  mercurius/lib/federation             |     100 |      100 |     100 |     100 |                   
>   compositionRules.js                 |     100 |      100 |     100 |     100 |                   
>  mercurius/lib/gateway                |   96.77 |     97.9 |     100 |   96.61 |                   
>   make-resolver.js                    |     100 |      100 |     100 |     100 |                   
>   request.js                          |     100 |      100 |     100 |     100 |                   
>   service-map.js                      |   90.91 |    91.18 |     100 |   90.91 | 92-97,141-147     
>  mercurius/tap-snapshots/test         |     100 |      100 |     100 |     100 |                   
>   errors.js.test.cjs                  |     100 |      100 |     100 |     100 |                   
>   federation.js.test.cjs              |     100 |      100 |     100 |     100 |                   
>   reply-decorator.js.test.cjs         |     100 |      100 |     100 |     100 |                   
>   routes.js.test.cjs                  |     100 |      100 |     100 |     100 |                   
>  mercurius/tap-snapshots/test/gateway |     100 |      100 |     100 |     100 |                   
>   remote-services.js.test.cjs         |     100 |      100 |     100 |     100 |                   
> --------------------------------------|---------|----------|---------|---------|-------------------
> ```

the gateway schema poll interval is failing periodically 😞 I've been re-runing the actions for them to pass

**Comment by mcollina** [comment]
@PabloSzx we should really refactor those tests. Maybe open a an issue or a PR?

**Comment by jonnydgreen** [comment]
Happy to help out with this if you need as the `remote-services.js` tests take a long time to run as well which would be great to sort at the same time!

**Comment by mcollina** [comment]
go for it Jonny!
